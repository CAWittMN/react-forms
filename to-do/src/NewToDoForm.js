import React, { useState } from "react";
import { v4 as uuid } from "uuid";
import "./NewToDoForm.css";

const NewToDoForm = ({ addToDo }) => {
  const INITIAL_STATE = { text: "" };
  const [formData, setFormData] = useState(INITIAL_STATE);

  const handleChange = (e) => {
    const { name, value } = e.target;
    setFormData((oldFormData) => ({
      ...oldFormData,
      [name]: value,
    }));
  };

  const handleSubmit = (e) => {
    e.preventDefault();
    addToDo({ ...formData, id: uuid() });
    setFormData(INITIAL_STATE);
  };

  return (
    <form className="NewToDoForm" onSubmit={handleSubmit}>
      <label htmlFor="text">New To Do:</label>
      <input
        id="text"
        type="text"
        name="text"
        value={formData.text}
        onChange={handleChange}
      />
      <button>Add To Do</button>
    </form>
  );
};

export default NewToDoForm;
