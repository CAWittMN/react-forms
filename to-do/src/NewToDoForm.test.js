import NewToDoForm from "./NewToDoForm";
import { render, fireEvent } from "@testing-library/react";

jest.mock("uuid", () => {
  return {
    v4: jest.fn(() => 1),
  };
});

it("renders without crashing", () => {
  render(<NewToDoForm />);
});

it("matches snapshot", () => {
  const { asFragment } = render(<NewToDoForm />);
  expect(asFragment()).toMatchSnapshot();
});

it("can update form data", () => {
  const { getByLabelText } = render(<NewToDoForm />);

  const taskInput = getByLabelText("New To Do:");

  fireEvent.change(taskInput, { target: { value: "do laundry" } });

  expect(taskInput).toHaveValue("do laundry");
});

it("can submit form", () => {
  const addToDo = jest.fn();
  const { getByLabelText, queryByText } = render(
    <NewToDoForm addToDo={addToDo} />
  );

  // fill out the form
  const taskInput = getByLabelText("New To Do:");
  fireEvent.change(taskInput, { target: { value: "do laundry" } });

  // submit form
  const submitBtn = queryByText("Add To Do");
  fireEvent.click(submitBtn);

  expect(addToDo).toHaveBeenCalledWith({
    text: "do laundry",
  });
});
