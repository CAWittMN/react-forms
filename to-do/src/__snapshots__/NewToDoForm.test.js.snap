// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <form
    class="NewToDoForm"
  >
    <label
      for="text"
    >
      New To Do:
    </label>
    <input
      id="text"
      name="text"
      type="text"
      value=""
    />
    <button>
      Add To Do
    </button>
  </form>
</DocumentFragment>
`;
